examples/subplots_axes_and_figures/subplots_demo.py

```diff
@@ -1,103 +1,190 @@
-
-f, axarr = plt.subplots(3, sharex=True, sharey=True)
-f.suptitle('Sharing both axes')
-axarr[0].plot(x, y)
-axarr[1].scatter(x, y)
-axarr[2].scatter(x, 2 * y ** 2 - 1, color='r')
-# Bring subplots close to each other.
-f.subplots_adjust(hspace=0)
+# Setting *sharex* or *sharey* to ``True`` enables global sharing across the
+# whole grid, i.e. also the y-axes of vertically stacked subplots have the
+# same scale when using ``sharey=True``.
+
+fig, axs = plt.subplots(3, sharex=True, sharey=True)
+fig.suptitle('Sharing both axes')
+axs[0].plot(x, y ** 2)
+axs[1].plot(x, 0.3 * y, 'o')
+axs[2].plot(x, y, '+')
+
+###############################################################################
+# For subplots that are sharing axes one set of tick labels is enough. Tick
+# labels of inner Axes are automatically removed by *sharex* and *sharey*.
+# Still there remains an unused empty space between the subplots.
+#
+# The parameter *gridspec_kw* of `.pyplot.subplots` controls the grid
+# properties (see also `.GridSpec`). For example, we can reduce the height
+# between vertical subplots using ``gridspec_kw={'hspace': 0}``.
+#
+# `.label_outer` is a handy method to remove labels and ticks from subplots
+# that are not at the edge of the grid.
+
+fig, axs = plt.subplots(3, sharex=True, sharey=True, gridspec_kw={'hspace': 0})
+fig.suptitle('Sharing both axes')
+axs[0].plot(x, y ** 2)
+axs[1].plot(x, 0.3 * y, 'o')
+axs[2].plot(x, y, '+')
+
 # Hide x labels and tick labels for all but bottom plot.
-for ax in axarr:
+for ax in axs:
     ax.label_outer()
 
 ###############################################################################
-# Row and column sharing
+# Apart from ``True`` and ``False``, both *sharex* and *sharey* accept the
+# values 'row' and 'col' to share the values only per row or column.
 
-f, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2, 2, sharex='col', sharey='row')
-f.suptitle('Sharing x per column, y per row')
+fig, axs = plt.subplots(2, 2, sharex='col', sharey='row',
+                        gridspec_kw={'hspace': 0, 'wspace': 0})
+(ax1, ax2), (ax3, ax4) = axs
+fig.suptitle('Sharing x per column, y per row')
 ax1.plot(x, y)
-ax2.scatter(x, y)
-ax3.scatter(x, 2 * y ** 2 - 1, color='r')
-ax4.plot(x, 2 * y ** 2 - 1, color='r')
+ax2.plot(x, y**2, 'tab:orange')
+ax3.plot(x + 1, -y, 'tab:green')
+ax4.plot(x + 2, -y**2, 'tab:red')
 
-###############################################################################
-# Four axes, returned as a 2-d array
-
-f, axarr = plt.subplots(2, 2)
-axarr[0, 0].plot(x, y)
-axarr[0, 0].set_title('Axis [0,0]')
-axarr[0, 1].scatter(x, y)
-axarr[0, 1].set_title('Axis [0,1]')
-axarr[1, 0].plot(x, y ** 2)
-axarr[1, 0].set_title('Axis [1,0]')
-axarr[1, 1].scatter(x, y ** 2)
-axarr[1, 1].set_title('Axis [1,1]')
-for ax in axarr.flat:
-    ax.set(xlabel='x-label', ylabel='y-label')
-# Hide x labels and tick labels for top plots and y ticks for right plots.
-for ax in axarr.flat:
+for ax in axs:
```

Comment:
`axs` here is an 2 by 2 array, so `ax` is also an array. For this to work, you need to flatten `axs`.

Reply:
I went ahead and fixed it. Hope that's fine!